File: gateway/src/models/commsec/commsec_contract.go
package commsec

import "github.com/google/uuid"

// ECDHEHandshakeInitRequest 表示 ECDHE 握手初始化请求。
type ECDHEHandshakeInitRequest struct {
	Initiator ServiceKeyOwner
	Responder ServiceKeyOwner

	InitiatorKeyID string

	SupportedKeyExchanges []KeyExchangeAlgorithm
	SupportedSignatures   []SignatureAlgorithm
	SupportedCipherSuites []CipherSuite

	Binding SecureChannelBinding
	TTLSec  int64
}

// ECDHEHandshakeInitResult 表示握手初始化结果。
type ECDHEHandshakeInitResult struct {
	Handshake ECDHEHandshakeRecord

	SelectedKeyExchange KeyExchangeAlgorithm
	SelectedSignature   SignatureAlgorithm
	SelectedCipherSuite CipherSuite
}

// ECDHEHandshakeCompleteRequest 表示 ECDHE 握手完成请求。
type ECDHEHandshakeCompleteRequest struct {
	HandshakeID uuid.UUID

	ResponderEphemeralPublicKey string
	ResponderSignature          string
	ResponderNonce              string
}

// ECDHEHandshakeCompleteResult 表示 ECDHE 握手完成结果。
type ECDHEHandshakeCompleteResult struct {
	Handshake ECDHEHandshakeRecord
	Channel   *SecureChannelSession
}

// SecureChannelUpsertRequest 表示建立或更新通道缓存请求。
type SecureChannelUpsertRequest struct {
	HandshakeID uuid.UUID
	Binding     SecureChannelBinding

	Source ServiceKeyOwner
	Target ServiceKeyOwner

	LocalKeyID string
	PeerKeyID  string

	CipherSuite   CipherSuite
	DerivedKeyRef string
	TTLSec        int64
}

// SecureChannelQuery 表示通道查询条件。
type SecureChannelQuery struct {
	ChannelID uuid.UUID
	Binding   SecureChannelBinding

	SourceServiceID string
	TargetServiceID string
}

// SecureChannelRevokeRequest 表示撤销通道请求。
type SecureChannelRevokeRequest struct {
	ChannelID uuid.UUID

	Binding   SecureChannelBinding
	Reason    string
	RevokedBy string
}

// SecureChannelEnsureRequest 表示“主动通信方”在出站前确保通道可用的请求。
type SecureChannelEnsureRequest struct {
	Query SecureChannelQuery

	HandshakeInit *ECDHEHandshakeInitRequest

	RequireActive    bool
	ForceReHandshake bool
}

// EncryptForChannelRequest 表示基于指定安全通道进行加密的请求。
type EncryptForChannelRequest struct {
	ChannelID uuid.UUID
	Payload   string

	AdditionalData map[string]string
	SequenceHint   uint64
}

// EncryptedPayload 表示应用层加密后的载荷和元数据。
type EncryptedPayload struct {
	CipherText string
	Meta       EncryptedMessageMeta
}

// DecryptFromChannelRequest 表示基于指定安全通道进行解密的请求。
type DecryptFromChannelRequest struct {
	ChannelID  uuid.UUID
	CipherText string
	Meta       EncryptedMessageMeta
}

// DecryptedPayload 表示解密后的明文以及更新后的通道状态信息。
type DecryptedPayload struct {
	Payload         string
	ChannelID       uuid.UUID
	UpdatedSequence uint64
}

File: data_worker/src/models/commsec/commsec_contract.py
from msgspec import Struct

from src.models.commsec.commsec import (
    CipherSuite,
    ECDHEHandshakeRecord,
    KeyExchangeAlgorithm,
    SecureChannelBinding,
    SecureChannelSession,
    ServiceKeyOwner,
    SignatureAlgorithm,
)


class ECDHEHandshakeInitRequest(Struct, kw_only=True):
    """表示ECDHE握手初始化请求参数。"""

    initiator: ServiceKeyOwner
    responder: ServiceKeyOwner

    initiator_key_id: str

    supported_key_exchanges: list[KeyExchangeAlgorithm]
    supported_signatures: list[SignatureAlgorithm]
    supported_cipher_suites: list[CipherSuite]

    binding: SecureChannelBinding
    ttl_sec: int


class ECDHEHandshakeInitResult(Struct, kw_only=True):
    """表示ECDHE握手初始化结果，包括握手记录和选定的算法等信息。"""

    handshake: ECDHEHandshakeRecord

    selected_key_exchange: KeyExchangeAlgorithm
    selected_signature: SignatureAlgorithm
    selected_cipher_suite: CipherSuite


class ECDHEHandshakeCompleteRequest(Struct, kw_only=True):
    """表示ECDHE握手完成请求参数。"""

    handshake_id: str

    responder_ephemeral_public_key: str
    responder_signature: str
    responder_nonce: str


class ECDHEHandshakeCompleteResult(Struct, kw_only=True):
    """表示ECDHE握手完成结果，包括握手记录和安全通道会话。"""

    handshake: ECDHEHandshakeRecord
    channel: SecureChannelSession | None = None


class SecureChannelUpsertRequest(Struct, kw_only=True):
    """表示建立或更新通道缓存请求。"""

    handshake_id: str
    binding: SecureChannelBinding

    source: ServiceKeyOwner
    target: ServiceKeyOwner

    local_key_id: str
    peer_key_id: str

    cipher_suite: CipherSuite
    derived_key_ref: str
    ttl_sec: int


class SecureChannelQuery(Struct, kw_only=True):
    """表示查询安全通道请求。"""

    channel_id: str
    binding: SecureChannelBinding

    source_service_id: str
    target_service_id: str


class SecureChannelRevokeRequest(Struct, kw_only=True):
    """表示撤销安全通道请求。"""

    channel_id: str

    binding: SecureChannelBinding
    reason: str
    revoked_by: str
